java-checks/src/main/java/org/sonar/java/checks/aws/AwsLambdaSyncCallCheck.java

```diff
@@ -0,0 +1,191 @@
+
+        if (isParameter(invokeRequest.symbol())
+          || localUsages.stream().anyMatch(lu -> isArgumentToACall(lu) || statementSetsAsyncCall(lu))
+          || declarationSetsAsyncCall(invokeRequest)) {
+          return Optional.empty();
+        }
+
+        if (hasLocalVarDeclaration(invokeRequest)) {
+          return Optional.of(MESSAGE);
+        }
+        return Optional.empty();
+      } else {
+        return Optional.empty();
+      }
+    }
+
+    private static boolean isArgumentToACall(IdentifierTree invokeRequest) {
+      return invokeRequest.parent() != null && invokeRequest.parent().is(Tree.Kind.ARGUMENTS);
+    }
+
+    private static boolean hasLocalVarDeclaration(IdentifierTree invokeRequest) {
+      Tree declaration = invokeRequest.symbol().declaration();
+      return (declaration != null && declaration.is(Tree.Kind.VARIABLE) && isLocalVariable(((VariableTree) declaration).symbol()));
+    }
+
+    /**
+     * Returns true if the statement starting at 'tree' sets the InvocationType object
+     * to a configuration that leads to async lambda calls. The statement could be a sequence of calls.
+     *
+     * @param tree the statement starting point
+     * @return true if statement leads lambda calls to be async
+     */
+    private static boolean statementSetsAsyncCall(Tree tree) {
+      MethodInvocationTree methodCall = findMethodInvocationTreeAncestor(tree);
+      if (methodCall != null) {
+        if (setsInvocationTypeToAsync(methodCall)) {
+          return true;
+        } else {
+          return statementSetsAsyncCall(methodCall);
+        }
+      }
+      return false;
+    }
+
+    private static MethodInvocationTree findMethodInvocationTreeAncestor(Tree tree) {
+      Tree parent = tree.parent();
+      while (parent != null) {
+        if (parent.is(Tree.Kind.METHOD_INVOCATION)) {
+          return (MethodInvocationTree) parent;
+        }
+        parent = parent.parent();
+      }
+      return null;
+    }
+
+    /**
+     * Returns true if the declaration of the 'invokeRequest' variable leads to async lambda calls.
+     *
+     * @param invokeRequest the 'invokeRequest' varible being declared
+     * @return true if the declaration of the 'invokeRequest' variable leads to async lambda calls
+     */
+    private static boolean declarationSetsAsyncCall(IdentifierTree invokeRequest) {
+      Tree declaration = invokeRequest.symbol().declaration();
+      AsyncInvocationTypeSetterFinder asyncSetterVisitor = new AsyncInvocationTypeSetterFinder();
+      // There has to be a declaration (if code compiled)
+      declaration.accept(asyncSetterVisitor);
+      return asyncSetterVisitor.found();
+    }
+
+    private static final class AsyncInvocationTypeSetterFinder extends BaseTreeVisitor {
+      private boolean found = false;
+
+      @Override
+      public void visitMethodInvocation(MethodInvocationTree methodCall) {
+        found = setsInvocationTypeToAsync(methodCall) || found;
```

Comment:
Would turning around the conditions be more efficient (i.e. once `found == true`, we wouldn't have to call `setsInvocationTypeToAsync` anymore)?